docs/source/advanced/splitters.rst

```diff
@@ -0,0 +1,168 @@
+.. _splitters:
+
+Splitters
```

Comment:
This is probably going to be annoying, but I don't think the name `Splitters` is great. The new `Processors` and `Buffers` create an immediate impression of what they might do before I even read a word of documentation because they inherit from predefined meanings in CS that give good clues. For me data splitting doesn't equal batching. For instance, in the ML world data splits are ways of permanently separating data from one another for purposes of training and test evaluation.

Batch terminology is also used throughout this documentation which might be a further hint that something around batches might be more appropriate. 

I also see that using  `Link(..., batching=batch_by_name)`or `Link(..., batches=batch_by_name)` doesn't exactly flow either.

 One signature I've seen used in other packages that might be appropriate is something along the lines of one of these options below:
- `Link(..., batch_fns=batch_by_name)`
- `Link(..., batching_fns=batch_by_name)`
-  `Link(..., batch_functions=batch_by_name)`

My preference would be towards the first one.

Reply:
Ok, interesting! I'm happy to discus this further.

Replying to your observations:
 
> For me data splitting doesn't equal batching. For instance, in the ML world data splits are ways of permanently separating data from one another for purposes of training and test evaluation.

Yeah, fair enough - but to me also 'buffer' and 'processor' have distinct (and varied) meanings depending on the context. What I think we're doing here is assigning a meaning that suit our use case, rather than trying to find words that purely and uniquely describe it. If we say that processor, splitter and buffer mean 'this and that' in Databay, then in my opinion it is a freedom we have - given that while not entirely precise, these aren't completely unrelated and misleading - merely somehow ambiguous. While it would be awesome to have a very specific word that describes 'the thing that divides into batches', one such doesn't exist and we'll have to compromise, just like we actually do with 'buffer' and especially 'processor'.

I want to hear your side of the argument too though. I've looked up the synonyms and things like `divider`, `divisor`, `separator`, `breaker`, `slicer` come up. Not sure about any of these yet, but I'd say these alternatives work better than appending an acronym behind `batch_`, or `batch_functions`.

Reply:
Yeah this is an interesting one for sure. 

> Yeah, fair enough - but to me also 'buffer' and 'processor' have distinct (and varied) meanings depending on the context.

Totally agree with this. I think what I was trying to get across with these is that they do hold a general nugget of meaning in many developers minds which give a clue as to what the feature or class will be doing, they make API easier to interact with off the bat for new users. `Splitters` is for me new terminology, or at least I haven't come across it before in this way. Of course if a word doesn't exist to describe what it does then as you say we'll have to find a way around it. Out of those synonyms you provided `slicer` is the one that I liked. I did take another look at the documentation and noticed how prevalent the word batching is:

> A splitter is a :any:`callable` function that accepts a list of batches and returns a list of batches.
> A list of batches is a two-dimensional list containing the records divided into sub-lists.
> Each of these sub-lists is called a batch.

It go me thinking that maybe `batcher` as a name might actually be appropriate here.

Reply:
Cool, I see your point man. In all fairness though:

> I think what I was trying to get across with these is that they do hold a general nugget of meaning in many developers minds which give a clue as to what the feature or class will be doing, they make API easier to interact with off the bat for new users. Splitters is for me new terminology, or at least I haven't come across it before in this way.

That's subjective, don't you think? Another developer could never have heard 'buffer' in this sense and claim that this is new terminology in this way too. Same for 'processors'. 

Maybe let's go with `batcher` though? While it doesn't seem to be an official word widely supported and it is a neologism - it is unambiguous.

Reply:
Yeah, it's certainly a very subjective thing. I think the point I was trying to make was that although it is certainly subjective if you were to look at the names as probabilities of terms developers are familiar with, `processor` and `buffer` would be higher than `splitter`, so with `batcher` maybe at least the terminology rings a bit more familiar with most developers. Of course, I'm making huge intuitive jumps as to what terminology developers are familiar with but that would be my guess.

Reply:
I can totally see where you're coming from - and yeah in such cases only just a subjective guess is also valuable. Let go with `batcher` then, I'm glad you brought this up as now I think I like it more than `splitter` too. Neither is perfect, but I think this is progress 👍